File: src/core/host_session.py
from __future__ import annotations

import os
import json
from typing import Literal, TypedDict

HostProvider = Literal["gemini", "codex", "claude"]


class HostBridgeConfig(TypedDict):
    command: str
    args: list[str]


def _normalize_flag(value: str | None) -> bool | None:
    if value is None:
        return None

    normalized = value.strip().lower()
    if normalized in {"1", "true", "yes", "on"}:
        return True
    if normalized in {"0", "false", "no", "off"}:
        return False
    return None


def detect_host_provider(env: dict[str, str] | None = None) -> HostProvider | None:
    current_env = env if env is not None else os.environ
    override = current_env.get("CORVUS_HOST_PROVIDER", "").strip().lower()
    if override in {"gemini", "codex", "claude"}:
        return override  # type: ignore[return-value]

    if current_env.get("GEMINI_CLI_ACTIVE") == "true":
        return "gemini"

    if current_env.get("CODEX_SHELL") == "1" or current_env.get("CODEX_THREAD_ID"):
        return "codex"

    return None


def is_host_session_active(env: dict[str, str] | None = None) -> bool:
    current_env = env if env is not None else os.environ
    override = _normalize_flag(current_env.get("CORVUS_HOST_SESSION_ACTIVE"))
    if override is not None:
        return override

    return detect_host_provider(current_env) is not None


def resolve_host_provider(
    env: dict[str, str] | None = None,
    fallback: HostProvider = "gemini",
) -> HostProvider | None:
    current_env = env if env is not None else os.environ
    override = _normalize_flag(current_env.get("CORVUS_HOST_SESSION_ACTIVE"))
    if override is False:
        return None

    provider = detect_host_provider(env)
    if provider is not None:
        return provider

    return fallback if is_host_session_active(env) else None


def _parse_bridge_args_json(raw: str | None, env_name: str) -> list[str]:
    if raw is None or not raw.strip():
        return ["{prompt}"]

    try:
        parsed = json.loads(raw)
    except json.JSONDecodeError as exc:
        raise RuntimeError(f"{env_name} must be valid JSON: {exc}") from exc

    if not isinstance(parsed, list) or not all(isinstance(entry, str) for entry in parsed):
        raise RuntimeError(f"{env_name} must be a JSON array of strings.")

    return [str(entry) for entry in parsed]


def _provider_bridge_env_names(provider: HostProvider) -> tuple[str, str]:
    prefix = f"CORVUS_{provider.upper()}_HOST_BRIDGE"
    return f"{prefix}_CMD", f"{prefix}_ARGS_JSON"


def resolve_configured_host_bridge(
    provider: HostProvider,
    env: dict[str, str] | None = None,
) -> HostBridgeConfig | None:
    current_env = env if env is not None else dict(os.environ)
    provider_cmd_key, provider_args_key = _provider_bridge_env_names(provider)
    provider_command = current_env.get(provider_cmd_key, "").strip()
    if provider_command:
        return {
            "command": provider_command,
            "args": _parse_bridge_args_json(current_env.get(provider_args_key), provider_args_key),
        }

    shared_command = current_env.get("CORVUS_HOST_BRIDGE_CMD", "").strip()
    if shared_command:
        return {
            "command": shared_command,
            "args": _parse_bridge_args_json(
                current_env.get("CORVUS_HOST_BRIDGE_ARGS_JSON"),
                "CORVUS_HOST_BRIDGE_ARGS_JSON",
            ),
        }

    return None


def expand_host_bridge_args(
    template: list[str],
    *,
    prompt: str,
    project_root: str,
    provider: HostProvider,
) -> list[str]:
    return [
        entry.replace("{prompt}", prompt)
        .replace("{project_root}", project_root)
        .replace("{provider}", provider)
        for entry in template
    ]


def get_host_bridge_configuration_hint(provider: HostProvider) -> str:
    provider_cmd_key, provider_args_key = _provider_bridge_env_names(provider)
    return (
        f"Set {provider_cmd_key} and {provider_args_key}, "
        "set CORVUS_HOST_BRIDGE_CMD and CORVUS_HOST_BRIDGE_ARGS_JSON, "
        "or supply an explicit host_session_runner."
    )

File: src/core/host_session.ts
export type HostProvider = 'gemini' | 'codex' | 'claude' | 'droid';
export type HostAttemptEvidence = Readonly<{
    requested_provider: HostProvider | null;
    actual_provider: HostProvider | null;
    requested_surface: string;
    actual_surface: string | null;
    execution_dispatched: boolean;
}>;
export type AugurySteeringMode = 'full' | 'lite';
export type HostSupportStatus =
    | 'supported'
    | 'native-session'
    | 'exec-bridge'
    | 'policy-only'
    | 'unsupported'
    | 'unknown';

export interface HostBridgeConfig {
    command: string;
    args: string[];
}

export interface HostDelegateBridgeConfig {
    command: string;
    args: string[];
}

export interface HostDelegatePollBridgeConfig {
    command: string;
    args: string[];
}

export type CapabilityExecutionMode =
    | 'agent-native'
    | 'kernel-backed'
    | 'policy-only'
    | 'unknown';

export type CapabilityOwnershipModel =
    | 'host-workflow'
    | 'kernel-primitive';

export type CapabilityKernelFallbackPolicy =
    | 'allowed'
    | 'forbidden';

export interface HostSkillActivationRequest {
    skill_id: string;
    role?: string;
    intent: string;
    project_root: string;
    target_paths?: string[];
    payload?: Record<string, unknown>;
    augury_contract?: Record<string, unknown>;
    augury_mode?: AugurySteeringMode;
    target_domain?: string;
    spoke_name?: string;
    requested_root?: string;
}

export interface AuguryLearningMetadata {
    [key: string]: unknown;
    schema_version: 1;
    steering_block_version: 2;
    steering_mode: AugurySteeringMode;
    corvus_standard_version: 1;
    optimizer_ready: true;
    optimizer_family: 'GEPA_DSPY';
    contract_hash: string;
    confidence?: number;
    confidence_source: 'explicit' | 'missing' | 'synthetic';
    route?: string;
    intent_category?: string;
    selection_tier?: string;
    selection_name?: string;
    expert_id?: string;
    expert_label?: string;
    council_candidates?: Array<{
        id: string;
        label: string;
        score: number;
        reason: string;
    }>;
    mimirs_well_count: number;
    mimirs_well_omitted_count: number;
    session_id?: string | null;
    planning_session_id?: string | null;
    designation_source?: string | null;
    prompt_surface?: string | null;
    bead_id?: string | null;
    weave_id?: string | null;
    result_status?: string | null;
    provider?: string | null;
    prompt_token_estimate?: number | null;
    target_domain?: string | null;
    spoke_name?: string | null;
    requested_root?: string | null;
}

export interface AuguryLearningEvent {
    schema_version: 1;
    event_version: 1;
    event_type: 'host_prompt';
    recorded_at: string;
    project_root: string;
    prompt_key?: string | null;
    prompt_surface?: string | null;
    steering_mode: AugurySteeringMode;
    contract_hash: string;
    confidence?: number;
    confidence_source: 'explicit' | 'missing' | 'synthetic';
    route?: string;
    intent_category?: string;
    selection_tier?: string;
    selection_name?: string;
    expert_id?: string;
    expert_label?: string;
    council_candidates?: AuguryLearningMetadata['council_candidates'];
    mimirs_well_count: number;
    mimirs_well_omitted_count: number;
    session_id?: string | null;
    planning_session_id?: string | null;
    designation_source?: string | null;
    provider?: string | null;
    target_domain?: string | null;
    spoke_name?: string | null;
    requested_root?: string | null;
    result_status?: string | null;
    transport_mode?: string | null;
    error?: string | null;
}

export interface AugurySteeringContext {
    mode?: AugurySteeringMode;
    project_root?: string;
    target_paths?: string[];
    target_domain?: string;
    spoke_name?: string;
    requested_root?: string;
}

function normalizeFlag(value: string | undefined): boolean | undefined {
    if (value === undefined) {
        return undefined;
    }

    const normalized = value.trim().toLowerCase();
    if (['1', 'true', 'yes', 'on'].includes(normalized)) {
        return true;
    }
    if (['0', 'false', 'no', 'off'].includes(normalized)) {
        return false;
    }
    return undefined;
}

export function detectHostProvider(env: NodeJS.ProcessEnv = process.env): HostProvider | null {
    const override = env.CORVUS_HOST_PROVIDER?.trim().toLowerCase();
    if (override === 'gemini' || override === 'codex' || override === 'claude' || override === 'droid') {
        return override as HostProvider;
    }

    if (env.CODEX_SHELL === '1' || Boolean(env.CODEX_THREAD_ID)) {
        return 'codex';
    }

    if (env.GEMINI_CLI_ACTIVE === 'true' || env.GEMINI_CLI === '1') {
        return 'gemini';
    }

    if (env.DROID_CLI_ACTIVE === 'true') {
        return 'droid';
    }

    return null;
}

export function isHostSessionActive(env: NodeJS.ProcessEnv = process.env): boolean {
    const override = normalizeFlag(env.CORVUS_HOST_SESSION_ACTIVE);
    if (override !== undefined) {
        return override;
    }

    return detectHostProvider(env) !== null;
}

export function isInteractiveHostSession(env: NodeJS.ProcessEnv = process.env): boolean {
    const override = normalizeFlag(env.CORVUS_HOST_SESSION_ACTIVE);
    if (override === false) {
        return false;
    }

    const provider = detectHostProvider(env);
    if (provider === 'gemini') {
        return env.GEMINI_CLI_ACTIVE === 'true' || env.GEMINI_CLI === '1';
    }
    if (provider === 'codex') {
        return env.CODEX_SHELL === '1';
    }
    if (provider === 'droid') {
        return env.DROID_CLI_ACTIVE === 'true';
    }
    return false;
}

export function resolveHostProvider(
    env: NodeJS.ProcessEnv = process.env,
): HostProvider | null {
    const override = normalizeFlag(env.CORVUS_HOST_SESSION_ACTIVE);
    if (override === false) {
        return null;
    }

    return detectHostProvider(env);
}

export function getHostProviderBanner(provider: HostProvider | null): string {
    if (provider === 'codex') {
        return ' ◤ CODEX CLI INTEGRATION ACTIVE ◢ ';
    }
    if (provider === 'claude') {
        return ' ◤ CLAUDE CLI INTEGRATION ACTIVE ◢ ';
    }
    if (provider === 'droid') {
        return ' ◤ DROID CLI INTEGRATION ACTIVE ◢ ';
    }
    if (provider === 'gemini') {
        return ' ◤ GEMINI CLI INTEGRATION ACTIVE ◢ ';
    }
    return ' ◤ HOST PROVIDER UNRESOLVED ◢ ';
}

export function getHostMindLabel(provider: HostProvider | null): string {
    if (provider === 'codex') {
        return 'OPENAI CODEX';
    }
    if (provider === 'claude') {
        return 'CLAUDE HOST';
    }
    if (provider === 'droid') {
        return 'DROID-CONTROL';
    }
    if (provider === 'gemini') {
        return 'GEMINI-3.1-PRO';
    }
    return 'HOST SESSION';
}

export function createHostAttemptEvidence(input: HostAttemptEvidence): HostAttemptEvidence {
    return Object.freeze({ ...input });
}

export function formatHostAttemptEvidence(evidence: HostAttemptEvidence): string {
    return [
        `requested_provider=${evidence.requested_provider ?? 'unresolved'}`,
        `actual_provider=${evidence.actual_provider ?? 'unresolved'}`,
        `requested_surface=${evidence.requested_surface}`,
        `actual_surface=${evidence.actual_surface ?? 'none'}`,
        `execution_dispatched=${String(evidence.execution_dispatched)}`,
    ].join(' ');
}

export * from './host_session_bridges.js';
export * from './host_session_augury.js';
